### DISCUSSION TITLE
Cannot start TeaVM-compiled JavaScript code from Web Worker using eval()

### DISCUSSION THREAD

**Discussion by WorldEditAxe** [konsoletyper/teavm#731] [Q&A] [answered]
Hello there, I'm attempting to create a self-contained JavaScript file that spawns workers of itself. My approach fetches the JavaScript file, `eval()`s it, and then calls `main()` (defined by TeaVM output JavaScript file). However, `main` remains undefined. Despite numerous attempts/approaches I've tried, I was unable to launch the worker script.
Here's my JavaScript code:
```js
fetch('%s').then(response=>response.text().then(script=>{globalThis.eval(script);main();}))
```
I've tried different dynamic code execution methods, but none of them worked. Is there something I'm doing wrong?

**Comment by konsoletyper** [comment, marked as answer]
Instead of eval, did you try using [importScript](https://developer.mozilla.org/en-US/docs/Web/API/WorkerGlobalScope/importScripts)? As for your code, `eval` creates a separate scope, and `main` is defined there, not in the scope that actually calls `main`. With `eval`-based approach it would look something like that:

```
fetch('%s').then(response=>response.text().then(script=>{globalThis.eval(script + ";main();");}))
```

**Comment by WorldEditAxe** [comment]
Thank you for your response! Your suggestion worked, but I was able to alternatively get it working by converting the JavaScript file into a Function, appending `\nmain();` to the end of it, and binding it to `globalThis`. I'll probably take and use your suggestion for the foreseeable future, seeing that it doesn't mangle the JS source and is more readable.
